speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp

```diff
@@ -35,31 +81,327 @@ inline std::string GetTimeAsString(time_t p_date)
+
+    // Save any tick
+    GET_RANDOM_TICKCOUNT;
+    fileDataStorage.Save(nullptr, nullptr, tickCount);
+    expected << std::to_string(tickCount) << "\n";
+    fileDataStorage.SaveDecisions(decisions);
+
+    if (p_storeDecisions)
+    {
+        WriteExpectedDecisions(decisions, expected, p_doSteer, p_doBrake, p_doAccel, p_doGear, p_doLights);
+    }
+    // Finish the buffer file
+    expected << "END";
+    fileDataStorage.Shutdown();
+
+    // Read the written file
+    std::ifstream reader(path);
+    std::string fileContents((std::istreambuf_iterator<char>(reader)), std::istreambuf_iterator<char>());
+    reader.close();
+
+    // Check contents
+    ASSERT_STREQ(fileContents.c_str(), expected.str().c_str());
+}
+
+/// @brief Test whether the FileDataStorage stores properly when decisions should be saved but none are made
+TEST(FileDataStorageTests, DecisionsNoneStored)
+{
+    TestDataStorageSaveDecisions(true, false, false, false, false, false);
+}
+
+/// @brief Test whether the FileDataStorage stores properly when decisions should be saved and all decisions are made
+TEST(FileDataStorageTests, DecisionsAllStored)
+{
+    TestDataStorageSaveDecisions(true, true, true, true, true, true);
+}
+
+/// @brief Test whether the FileDataStorage stores properly when decisions should not be saved and none are made
+TEST(FileDataStorageTests, DecisionsDisabledNoDecisions)
+{
+    TestDataStorageSaveDecisions(false, false, false, false, false, false);
+}
+
+/// @brief Test whether the FileDataStorage stores properly when decisions should not be saved but all decisions are made
+TEST(FileDataStorageTests, DecisionsDisabledAllDecisions)
+{
+    TestDataStorageSaveDecisions(false, true, true, true, true, true);
+}
+
+/// @brief Test whether the FileDataStorage stores properly for every combination of decision made
+TEST(FileDataStorageTests, DecisionsCombinations)
+{
+    bool alwaysTrue[1]{true};
+    bool booleans[2]{true, false};
+    PairWiseTest(TestDataStorageSaveDecisions, alwaysTrue, 1, booleans, 2, booleans, 2, booleans, 2, booleans, 2, booleans, 2);
+}
+
+/// @brief Test writing decisions after data to ensure the order is as expected
+TEST(FileDataStorageTests, DecisionsAfterData)
+{
+    Random random;
+    tDataToStore params = {
+        true,
+        true,
+        true,
+        true,
+        true};
+    FileDataStorage fileDataStorage;
+
+    // Create a string to save all intended random data to
+    std::stringstream expected;
+
+    // Initialise buffer file
+    GET_DUMMY_TIMES;
+    std::experimental::filesystem::path path = fileDataStorage.Initialize(params, DUMMY_INITIALISATION_PARAMETERS);
+    expected << DUMMY_INITIALISATION_FILE_ENTRIES << "GameState\nUserInput\n";
+
+    // Generate and write data
+    GET_RANDOM_TICKCOUNT;
+    GET_RANDOM_CAR;
+    GET_RANDOM_SITUATION;
+    fileDataStorage.Save(&car, &situation, tickCount);
+    expected << std::to_string(tickCount) << "\n";
```

Comment:
Change all '\n' characters with std::endl as enters are different on windows and linux